src/lib/opds.ts

```diff
@@ -0,0 +1,104 @@
+/**
+ * Generation of the OPDS file for the full set of tests.
+ * 
+  * 
+ *  @packageDocumentation
+ */
+
+import { TestData, Constants } from './types';
+
+/* ------------------------------------------------------------------------------------------------------ */
+/*                        Subset of OPDS as used here in Typescript types                                 */
+/* ------------------------------------------------------------------------------------------------------ */
+
+
+interface PublicationMetadata {
+    "@type": string;
+    identifier: string;
+    title: string;
+    author: string | string[];
+    description: string;
+    collection: string;
+    modified: string;
+    publisher: string
+}
+
+interface PublicationLink {
+    type: string;
+    rel: string;
+    href: string;
+}
+
+interface ImageLink {
+    type: string;
+    href: string;
+    height?: number;
+    width?: number;
+}
+
+interface Publication {
+    metadata: PublicationMetadata;
+    links: PublicationLink[];
+    images: ImageLink[];
+}
+
+export interface OPDS {
+    metadata: {
+        title: string;
+    };
+    links : PublicationLink[];
+    publications: Publication[];
+}
+
+export function create_opds(tests: TestData[]): OPDS {
+
+    /** All cover images are identical */
+    const images: ImageLink[] = [{
+        href   : `${Constants.OPDS_DIR_URL}/${Constants.DOC_OPDS_COVER_PNG}`,
+        type   : 'image/png',
+        height : 849,
+        width  : 600,
+    },{
+        href : `${Constants.OPDS_DIR_URL}/${Constants.DOC_OPDS_COVER_SVG}`,
+        type : 'image/svg',
+    }]
+
+    const publications = tests.map((test: TestData): Publication => {
+        const links: PublicationLink[] = [{
+            type : `${Constants.EPUB_MEDIA_TYPE}`,
+            rel  : 'http://opds-spec.org/acquisition/open-access',
+            href : `${Constants.TEST_DOWNLOAD_URL_BASE}/${test.identifier}.epub`,
+        }]
+        const metadata: PublicationMetadata = {
+            "@type"     : 'http://schema.org/Book',
+            identifier  : `${Constants.TEST_URL_BASE}/${test.identifier}`,
+            title       : test.title,
+            author      : test.creators.length === 1 ? test.creators[0] : test.creators,
+            description : test.description,
+            collection  : test.coverage,
+            modified    : test.modified,
+            publisher   : 'W3C',
```

Comment:
Bonjour Ivan :)

Could you please add:
```
subject: ['epub-test'],
```

...this way, importing publications from this OPDS feed will automatically inherit the `epub-test` "tag" which will be super useful in Thorium (for example), to triage the test suite using search/filter/sort in our v2.0 bookshelf (when running the tests, we will manually add our own tags, e.g. `epub-test-fail` / `epub-test-pass`, `epub-test-todo`, etc.).

Note that the HTTP content-type header for the OPDS feed is incorrect:
* https://raw.githubusercontent.com/w3c/epub-tests/generate_opds/docs/opds/opds.json
...so instead we use this URL in Thorium:
* https://raw.githack.com/w3c/epub-tests/generate_opds/docs/opds/opds.json

Also note that internally we occasionally use a online "micro service" which conveniently visualises a Readium Web Pub Manifest or ODPS feed, and more crucially that runs JSON Schema validation (scroll to the very bottom to see the result):
* https://readium2.herokuapp.com/opds-v2-browse/https%3A%2F%2Fraw.githack.com%2Fw3c%2Fepub-tests%2Fgenerate_opds%2Fdocs%2Fopds%2Fopds.json

RWPM `subject`:
* https://github.com/readium/webpub-manifest/tree/master/contexts/default#subjects
JSON Schema:
* https://github.com/readium/webpub-manifest/blob/ee2491413fe33133c42569c61e9aa1a51deb0d52/schema/metadata.schema.json#L101-L103
* https://github.com/readium/webpub-manifest/blob/ee2491413fe33133c42569c61e9aa1a51deb0d52/schema/subject.schema.json

Concrete real-world use of `subject` at Feedbooks (note the `links` property which allows "clicking" on a subject tag in the user interface to fetch an OPDS feed URL that shows only publications with the selected subject):

* http://readium2.herokuapp.com/opds-v2-browse/https%3A%2F%2Fcatalog.feedbooks.com%2Fitem%2F2159345.json
* https://catalog.feedbooks.com/item/2159345.json

Reply:
> Note that the HTTP content-type header for the OPDS feed is incorrect:
> 
> * [`generate_opds`/docs/opds/opds.json (raw)](https://raw.githubusercontent.com/w3c/epub-tests/generate_opds/docs/opds/opds.json)
>   ...so instead we use this URL in Thorium:
> * [raw.githack.com/w3c/epub-tests/generate_opds/docs/opds/opds.json](https://raw.githack.com/w3c/epub-tests/generate_opds/docs/opds/opds.json)

Just to check... Isn't this also a problem with the fact that this file is not yet on the main branch; it will be there when the PR is merged. Or is there an entry in the HTTP header that you expect and that the raw github response does not provide?

Reply:
IIRC the HTTP content-type for JSON will be emitted correctly if/when serving files from gh-pages or another publishing branch in the GitHub repository. However with all other branches it is served as plain text, (or octet stream, i cannot remember and I am on my phone soncannot eailsiky check ;) ... thus the use of GitHack or other suitablr CDN.

Reply:
Note the previous comment: there is another json file that is on the main branch, with the URI https://w3c.github.io/epub-tests/drafts/config.json. The HTTP headers are:

```
HTTP/2 200 
server: GitHub.com
content-type: application/json; charset=utf-8
permissions-policy: interest-cohort=()
x-origin-cache: HIT
last-modified: Sat, 28 May 2022 17:12:37 GMT
access-control-allow-origin: *
etag: "62925805-291"
expires: Wed, 08 Jun 2022 06:30:48 GMT
cache-control: max-age=600
x-proxy-cache: MISS
x-github-request-id: 7FA0:9420:988E6F:A3D4D9:62A03FC0
accept-ranges: bytes
date: Wed, 08 Jun 2022 06:20:48 GMT
via: 1.1 varnish
age: 0
x-served-by: cache-fco2270023-FCO
x-cache: MISS
x-cache-hits: 0
x-timer: S1654669248.401139,VS0,VE104
vary: Accept-Encoding
x-fastly-request-id: f23f77ec52c3771daed5e9a10407efa706325b35
content-length: 657
```

I would think that should work with thorium as well...

Reply:
that's because it is a GitHub.io URL (i.e. the source is a GitHub publishing branch, regardless of its symbolic name, equivalent to gh-pages).

Reply:
> that's because it is a GitHub.io URL (i.e. the source is a GitHub publishing branch, regardless of its symbolic name, equivalent to gh-pages).

Right. But once this PR is published, https://w3c.github.io/epub-tests/opds/opds.json will behave exactly the same way.

Ie, the overall link for the opds.json file should remain the same imho...

Reply:
B.t.w., I have done the changes for the subject, but still have to regenerate the files.

Reply:
> still have to regenerate the files

Also done...

Reply:
Thank you Ivan, it works great. See screenshots (ignoring the broken cover images, as previously discussed):
![Screenshot 2022-06-08 at 07 51 00](https://user-images.githubusercontent.com/625201/172551225-c04e45eb-b499-402c-908e-e4611402e850.jpg)
![Screenshot 2022-06-08 at 07 50 18](https://user-images.githubusercontent.com/625201/172551236-ca57a8b0-bacd-4162-8434-d8e2adbdf99c.jpg)
![Screenshot 2022-06-08 at 07 50 02](https://user-images.githubusercontent.com/625201/172551240-263aa5a9-ad57-4f63-9677-390c96ef9d7b.jpg)

Reply:
Now that the `epub-test` is automatically set in the local bookshelf when importing EPUBs from the OPDS feed, I can easily search/filter/sort in the "data grid" view, to show only the EPUB test suite, and/or to display only my `epub-test-pass` or `epub-test-fail` tags (or even `epub-test-question` etc. if I am unsure about a test):

![Screenshot 2022-06-08 at 07 56 19](https://user-images.githubusercontent.com/625201/172552064-bb549730-9ff2-450f-bd4a-63ac89493e6e.jpg)

Reply:
Wow. This all looks great!

I will add some words about the availability of this file in the general description. Can we expect additional review? @llemeurfr referred to Pierre possibly reviewing this... If so, I am happy to wait with merging.

Reply:
Well, the OPDS feed "works" on multiple levels:
* We can open it in Thorium, and download publications with automatic "tag" import (`subject` metadata). The broken cover image is not critical, and it will be fixed when merged anyway (if I understand correctly).
* This OPDS feed is in the v2 JSON format, which validates against the official JSON Schema (scroll all the way to the bottom of the page to see the result): https://readium2.herokuapp.com/opds-v2-browse/https%3A%2F%2Fraw.githack.com%2Fw3c%2Fepub-tests%2Fgenerate_opds%2Fdocs%2Fopds%2Fopds.json
* Alternative online validator link: https://www.jsonschemavalidator.net/s/GIauaQAy (JSON Schema source: https://raw.githubusercontent.com/opds-community/drafts/master/schema/feed.schema.json )

...so, as far I we are concerned, this PR can be merged. We will report back if there are issues.

Reply:
Daniel has carefully checked the file, therefore Pierre's review is not required anymore. Great work! that will be also useful for checking the test suite in many mobiles apps (ex. Aldiko).

Reply:
Ok! Merging now.